### DISCUSSION TITLE
Anemic models, service layers: benefits and drawbacks

### DISCUSSION THREAD

**Discussion by timkelty** [craftcms/cms#7893] [General]
Mostly just wanted to try out Discussions :)

Craft (and many other MVCs) seems to encourage robust services and thin models:
https://craftcms.com/docs/3.x/extend/services.html#what-are-services

> They have two jobs:
> 
> They contain most of your plugin’s business logic.
> They define your plugin’s API, which your plugin (and other plugins) can access.

While common, this is often called out as an anti-pattern: https://martinfowler.com/bliki/AnemicDomainModel.html

> The fundamental horror of this anti-pattern is that it's so contrary to the basic idea of object-oriented design; which is to combine data and process together.

That's not to say there is no place for a service layer, but Fowler & Evans argue it should be thin:

> This layer is kept thin. It does not contain business rules or knowledge, but only coordinates tasks and delegates work to collaborations of domain objects in the next layer down.

I tend to be on the side of minimizing services and having more behavior models, and of course, theres nothing preventing devs from writing their Craft plugins and modules this way, it just isn't what Craft does.

I'd love to hear arguments on either side!

**Comment by ostark** [comment]
Tim, thanks for starting the discussion on this topic!

Using examples and highlighting possible issues might help people translating these high-level design patterns into the real-world code.

### Common patterns I see with many Craft plugins:

* One or more services are registered as `components` of the Module / Plugin
* Services are used as a  "Repository" with methods like `deleteThingById(int $id)` or `getAllThings()`
* Services are coupled to the state of `Craft::$app` (request, currentSite)
* Services contain state
* Services contain tons of business logic 
* There is no further delegation or actual domain modeling using other classes
* Services are often called by event listener callbacks
* Services call other services (e.g.  `Plugin::$plugin->you->make()`)
* Dependency injection is not a thing
* Services have generic names and have many responsibilities (e.g.: `FooPlugin` with `FooService`)

**@timkelty Do you see this as well? What's wrong with it? 🙈**

I strongly believe, most people don't see the problem.

**Reply by timkelty** [reply]
@ostark yep to all those things.

I think documenting and talking about these things as we see them is a great first step though.

I feel like there's lots of following what others have done to get the job done, but I'd love to see more discussion around the _why_.

**Reply by ostark** [reply]
Another problem I see is tight coupling to the (yii) framework.
While it is convenient to access `Craft::$app` from any possible place in the codebase, it has downsides in the long term.

At some point, Craft will switch to a more modern framework. No matter if it's Yii3, Laravel 10, or a set of Symfony components, if the domain logic has too much knowledge about the framework, a complete rewrite is required. This is probably not an issue for a tiny plugin/module that contains less than 500 lines of code. But for bigger projects, it will require weeks.

Plugins are an important part of any CMS ecosystem. Helping developers to write better code, that is easier to maintain and migratable to a future version of Craft, should be a goal of P&T and the community. Otherwise, I fear the "WordPress Plugin Delemma" will apply to Craft in the future.

**Comment by angrybrad** [comment]
Yeah, when we make the next big framework upgrade, that'll be our chance to re-evaluate and update our architecture.

It'll obviously be easier with a framework that has taken those design considerations into account as well.

It's still a ways off but we're thinking about it and discussions are the perfect place for topics like this - appreciate it.
